db.js

```diff
@@ -731,13 +775,73 @@ exports.init = function (sbot, config) {
           done((err) => {
             // prettier-ignore
             if (err) return unlock(cb, clarify(err, 'reindexEncrypted() failed to force-flush indexes'))
+            indexingActive.set(indexingActive.value - 1)
             unlock(cb)
           })
         })
       )
     })
   }
 
+  function notYetZero(obz, fn, ...args) {
+    if (obz.value > 0) {
+      onceWhen(obz, (x) => x === 0, fn.bind(null, ...args))
+      return true
+    } else {
+      return false
+    }
+  }
+
+  function compact(cb) {
+    if (notYetZero(jitdb.indexingActive, compact, cb)) return
+    if (notYetZero(jitdb.queriesActive, compact, cb)) return
+    if (notYetZero(indexingActive, compact, cb)) return
+
+    fs.closeSync(fs.openSync(resetLevelPath(dir), 'w'))
+    fs.closeSync(fs.openSync(resetPrivatePath(dir), 'w'))
+    fs.closeSync(fs.openSync(reindexJitPath(dir), 'w'))
+    log.compact(function onLogCompacted(err) {
+      if (err) cb(clarify(err, 'ssb-db2 compact() failed with the log'))
+      else cb()
+    })
+  }
+
+  let compactStartOffset = null
+  log.compactionProgress((stats) => {
+    if (typeof stats.startOffset === 'number' && compactStartOffset === null) {
+      compactStartOffset = stats.startOffset
+    }
+
+    if (compacting.value !== !stats.done) compacting.set(!stats.done)
+
+    if (stats.done) {
+      if (stats.sizeDiff > 0) {
+        if (fs.existsSync(resetLevelPath(dir))) {
+          resetAllIndexes(() => {
+            rimraf.sync(resetLevelPath(dir))
+          })
+        }
+        if (fs.existsSync(resetPrivatePath(dir))) {
+          privateIndex.reset(() => {
+            rimraf.sync(resetPrivatePath(dir))
+          })
+        }
+        if (fs.existsSync(reindexJitPath(dir))) {
+          jitdb.reindex(compactStartOffset || 0, (err) => {
+            if (err) console.error('ssb-db2 reindex jitdb after compact', err)
+            rimraf.sync(reindexJitPath(dir))
+          })
+        }
+        compactStartOffset = null
+      } else {
+        rimraf.sync(resetLevelPath(dir))
+        rimraf.sync(resetPrivatePath(dir))
+        rimraf.sync(reindexJitPath(dir))
+        restartUpdateIndexes()
```

Comment:
This `restartUpdateIndexes()` addresses the following case:

- `log.compact()` was started immediately, before `updateIndexes()` could start
- Crash
- Resume
- compaction is resumed automatically by AAOL
- `updateIndexes()` is cancelled because compaction is ongoing
- compaction is done, but `sizeDiff === 0`
- **At this point, leveldb indexes are outdated** (they haven't had the chance to build at all!) but there is no `updateIndexes()` ongoing, so we have to kickstart it somehow

Reply:
I was looking at this to make sure that it works correctly in a crash and I think it does because:

You don't need to call reset on indexes because if that was needed (sizeDiff > 0) then we would already know because this info is persisted in AAOL and would not disappear in a crash.

In other words: 👍